gax/src/main/java/com/google/api/gax/rpc/ReframingResponseObserver.java

```diff
@@ -0,0 +1,368 @@
+          }
+        });
+
+    hasStarted = true;
+
+    if (autoFlowControl) {
+      numPending = Integer.MAX_VALUE;
+      deliver();
+    }
+  }
+
+  /**
+   * Request n responses to be delivered to the outer/downstream {@link
+   * ResponseObserver#onResponse(Object)}. This method might synchronously deliver the messages if
+   * they have already been buffered. Or it will deliver them asynchronously if they need to be
+   * requested from upstream.
+   *
+   * @param count The maximum number of responses to deliver
+   */
+  private void onRequest(int count) {
+    Preconditions.checkState(!autoFlowControl, "Auto flow control enabled");
+    Preconditions.checkArgument(count > 0, "Count must be > 0");
+
+    synchronized (lock) {
+      int maxCount = Integer.MAX_VALUE - numPending;
+      count = Math.min(maxCount, count);
+
+      numPending += count;
+    }
+
+    deliver();
+  }
+
+  /**
+   * Cancels the stream and notifies the downstream {@link ResponseObserver#onError(Throwable)}.
+   * This method can be called multiple times, but only the first time has any effect. Please note
+   * that there is a race condition between cancellation and the stream completing normally. Please
+   * note that you can only specify a message or a cause, not both.
+   */
+  private void onCancel() {
+    innerController.cancel();
+
+    synchronized (lock) {
+      deferError = false;
+      error = new CancellationException("User cancelled stream");
+    }
+
+    deliver();
+  }
+
+  /**
+   * Process a new response from inner/upstream callable. The message will be fed to the reframer
+   * and the output will be delivered to the downstream {@link ResponseObserver}.
+   *
+   * <p>If the delivery loop is stopped, this will restart it.
+   */
+  @Override
+  public void onResponse(InnerT response) {
+    synchronized (lock) {
+      awaitingInner = false;
+      reframer.push(response);
+    }
+    deliver();
+  }
+
+  /**
+   * Process inner/upstream callable's onError notification. This will be queued to be delivered
+   * after the delegate exhausts its buffers.
+   *
+   * <p>If the delivery loop is stopped, this will restart it.
+   */
+  @Override
+  public void onError(Throwable t) {
+    synchronized (lock) {
+      if (error == null) {
+        error = t;
+      }
+      closeOnDone = true;
+    }
+    deliver();
+  }
+
+  /**
+   * Process inner/upstream callable's onComplete notification. This will be queued to be delivered
+   * after the delegate exhausts its buffers.
+   *
+   * <p>If the delivery loop is stopped, this will restart it.
+   */
+  @Override
+  public void onComplete() {
+    synchronized (lock) {
+      closeOnDone = true;
+    }
+    deliver();
+  }
+
+  /**
+   * The message pump that interacts with the {@link this#outerResponseObserver}. All interactions
+   * with outer {@link ResponseObserver} will be serialized via this method.
+   */
+  /** Tries to kick off the delivery loop, wrapping it in error handling. */
```

Comment:
I don't think this comment belongs here?

Reply:
oops, rebase failure :(